File: backend/src/models/Book.ts
import mongoose, { Schema, Document, Types } from 'mongoose';
import { sanitizePageText } from '../lib/sanitize.js';

export type BookStatus =
  | 'uploading'
  | 'splitting_pages'
  | 'extracting_cover'
  | 'reading_title'
  | 'ocr_processing'
  | 'detecting_chapters'
  | 'awaiting_chapter_review'
  | 'generating_audio'
  | 'complete'
  | 'error';

export type AudioStatus = 'pending' | 'stale' | 'generating' | 'complete' | 'error';

// One audio chunk for a single sentence, in one voice. Kept on disk so a typo
// fix only re-renders this segment instead of the whole chapter.
export interface ISegment {
  sentenceId: Types.ObjectId;
  audioPath?: string;
  durationSecs?: number;
  audioStatus: AudioStatus;
  audioError?: string;
}

export interface IVoiceTrack {
  voice: string;
  audioPath?: string;
  audioDurationSecs?: number;
  audioStatus: AudioStatus;
  audioError?: string;
  segments: Types.DocumentArray<ISegment & Document>;
}

// An editable, speech-ready sentence — the source of truth for chapter audio.
// Shared across voices (text is the same); each voice renders its own segment.
export interface ISentence {
  _id: Types.ObjectId;
  order: number;
  // The speech-ready text that is actually synthesized (references/acronyms
  // expanded). `display` is the original reviewed text shown to the reader.
  text: string;
  display?: string;
}

export interface IChapter {
  _id: Types.ObjectId;
  title: string;
  startPage: number;
  startChar: number;
  sentences: Types.DocumentArray<ISentence & Document>;
  tracks: Types.DocumentArray<IVoiceTrack & Document>;
}

export interface IOcrPage {
  page: number;
  text: string;
  // The OCR'd, reviewed text rewritten for speech (references/acronyms expanded).
  // What actually gets read; kept alongside `text` so the review UI can diff them.
  readText?: string;
  language: string;
  status: 'pending' | 'processing' | 'complete' | 'error';
  error?: string;
}

export interface IBook extends Document {
  name: string;
  status: BookStatus;
  // Book-wide language (ISO 639-1), detected once from the summary page. Used for
  // speech normalization in place of per-page detection.
  language?: string;
  folderPath: string;
  filePath: string;
  coverImagePath?: string;
  // Pages holding the table of contents / index. The TOC is read from each and
  // the entries merged, so a contents list spanning several pages is supported.
  summaryPages: number[];
  coverPage: number;
  firstPage: number;
  lastPage: number;
  totalPages: number;
  voices: string[];
  chapters: Types.DocumentArray<IChapter & Document>;
  ocrPages: IOcrPage[];
  progress: { current: number; total: number; message: string };
  errorMessage?: string;
  deleted: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export function trackForVoice(
  chapter: IChapter,
  voice: string
): (IVoiceTrack & Document) | undefined {
  return chapter.tracks.find(t => t.voice === voice);
}

export function freshTracks(voices: string[]) {
  return voices.map(voice => ({ voice, audioStatus: 'pending' as AudioStatus, segments: [] as ISegment[] }));
}

// A track's status derived from its segments (the assembled-chapter readiness).
export function deriveTrackStatus(segments: { audioStatus: AudioStatus }[]): AudioStatus {
  if (segments.length === 0) return 'pending';
  if (segments.some(s => s.audioStatus === 'generating')) return 'generating';
  if (segments.some(s => s.audioStatus === 'error')) return 'error';
  if (segments.some(s => s.audioStatus === 'stale')) return 'stale';
  if (segments.every(s => s.audioStatus === 'complete')) return 'complete';
  return 'pending';
}

// Chapters trimmed for the wire: sentences and per-segment data stay server-side
// (the editor fetches them on demand) to keep sync/update payloads small.
export function serializeChaptersForClient(chapters: IBook['chapters']) {
  return chapters.map(c => ({
    _id: c._id,
    title: c.title,
    startPage: c.startPage,
    startChar: c.startChar,
    tracks: c.tracks.map(t => ({
      voice: t.voice,
      audioPath: t.audioPath,
      audioDurationSecs: t.audioDurationSecs,
      audioStatus: t.audioStatus,
      audioError: t.audioError,
    })),
  }));
}

const SegmentSchema = new Schema<ISegment>(
  {
    sentenceId: { type: Schema.Types.ObjectId, required: true },
    audioPath: { type: String },
    durationSecs: { type: Number },
    audioStatus: {
      type: String,
      enum: ['pending', 'stale', 'generating', 'complete', 'error'],
      default: 'pending',
    },
    audioError: { type: String },
  },
  { _id: false }
);

const VoiceTrackSchema = new Schema<IVoiceTrack>(
  {
    voice: { type: String, required: true },
    audioPath: { type: String },
    audioDurationSecs: { type: Number },
    audioStatus: {
      type: String,
      enum: ['pending', 'stale', 'generating', 'complete', 'error'],
      default: 'pending',
    },
    audioError: { type: String },
    segments: { type: [SegmentSchema], default: [] },
  },
  { _id: false }
);

const SentenceSchema = new Schema<ISentence>({
  order: { type: Number, required: true },
  text: { type: String, default: '' },
  display: { type: String, default: '' },
});

const ChapterSchema = new Schema<IChapter>({
  title: { type: String, required: true },
  startPage: { type: Number, required: true },
  startChar: { type: Number, default: 0 },
  sentences: { type: [SentenceSchema], default: [] },
  tracks: { type: [VoiceTrackSchema], default: [] },
});

const OcrPageSchema = new Schema<IOcrPage>({
  page: { type: Number, required: true },
  text: { type: String, default: '' },
  readText: { type: String },
  language: { type: String, default: 'unknown' },
  status: {
    type: String,
    enum: ['pending', 'processing', 'complete', 'error'],
    default: 'pending',
  },
  error: { type: String },
});

const BookSchema = new Schema<IBook>(
  {
    name: { type: String, default: '' },
    status: {
      type: String,
      enum: [
        'uploading', 'splitting_pages', 'extracting_cover', 'reading_title',
        'ocr_processing', 'detecting_chapters', 'awaiting_chapter_review',
        'generating_audio', 'complete', 'error',
      ],
      default: 'uploading',
    },
    language: { type: String },
    folderPath: { type: String, required: true },
    filePath: { type: String, required: true },
    coverImagePath: { type: String },
    summaryPages: { type: [Number], required: true },
    coverPage: { type: Number, required: true },
    firstPage: { type: Number, required: true },
    lastPage: { type: Number, required: true },
    totalPages: { type: Number, default: 0 },
    chapters: [ChapterSchema],
    ocrPages: [OcrPageSchema],
    progress: {
      current: { type: Number, default: 0 },
      total: { type: Number, default: 0 },
      message: { type: String, default: '' },
    },
    errorMessage: { type: String },
    voices: { type: [String], default: ['chatterbox:pt-BR-FranciscaNeural'] },
    deleted: { type: Boolean, default: false },
  },
  { timestamps: true }
);

export const Book = mongoose.model<IBook>('Book', BookSchema);

export async function migrateLegacyVoices(): Promise<void> {
  const col = mongoose.connection.collection('books');
  const cursor = col.find({ voices: { $exists: false } });

  for await (const b of cursor) {
    const voice: string = b.voice || 'chatterbox:pt-BR-FranciscaNeural';
    const chapters = (b.chapters || []).map((ch: Record<string, unknown>) => {
      if (Array.isArray(ch.tracks)) return ch;
      return {
        ...ch,
        tracks: [{
          voice,
          audioPath: ch.audioPath,
          audioDurationSecs: ch.audioDurationSecs,
          audioStatus: ch.audioStatus || 'pending',
        }],
      };
    });

    await col.updateOne(
      { _id: b._id },
      { $set: { voices: [voice], chapters }, $unset: { voice: '' } }
    );
  }
}

// Older books stored a single `summaryPage`; fold it into the new `summaryPages`
// array so multi-page contents lists are supported uniformly.
export async function migrateSummaryPages(): Promise<void> {
  const col = mongoose.connection.collection('books');
  const cursor = col.find({ summaryPages: { $exists: false } });

  for await (const b of cursor) {
    const page = typeof b.summaryPage === 'number' ? b.summaryPage : 0;
    await col.updateOne(
      { _id: b._id },
      { $set: { summaryPages: page > 0 ? [page] : [] }, $unset: { summaryPage: '' } }
    );
  }
}

export async function migrateSanitizeOcrText(): Promise<void> {
  const col = mongoose.connection.collection('books');
  const cursor = col.find({ 'ocrPages.text': { $regex: /^\s*\{/ } });

  for await (const b of cursor) {
    const ocrPages = (b.ocrPages || []).map((p: Record<string, unknown>) => ({
      ...p,
      text: sanitizePageText(p.text as string),
    }));
    await col.updateOne({ _id: b._id }, { $set: { ocrPages } });
  }
}

File: backend-go/internal/model/book.go
package model

import (
	"go.mongodb.org/mongo-driver/v2/bson"

	"github.com/kevynsax/book-reader/backend-go/internal/lib/sanitize"
)

type BookStatus string

const (
	StatusUploading             BookStatus = "uploading"
	StatusSplittingPages        BookStatus = "splitting_pages"
	StatusExtractingCover       BookStatus = "extracting_cover"
	StatusReadingTitle          BookStatus = "reading_title"
	StatusOcrProcessing         BookStatus = "ocr_processing"
	StatusDetectingChapters     BookStatus = "detecting_chapters"
	StatusAwaitingChapterReview BookStatus = "awaiting_chapter_review"
	StatusGeneratingAudio       BookStatus = "generating_audio"
	StatusComplete              BookStatus = "complete"
	StatusError                 BookStatus = "error"
)

type AudioStatus string

const (
	AudioPending    AudioStatus = "pending"
	AudioStale      AudioStatus = "stale"
	AudioGenerating AudioStatus = "generating"
	AudioComplete   AudioStatus = "complete"
	AudioError      AudioStatus = "error"
)

type OcrStatus string

const (
	OcrPending    OcrStatus = "pending"
	OcrProcessing OcrStatus = "processing"
	OcrComplete   OcrStatus = "complete"
	OcrErr        OcrStatus = "error"
)

// Segment is one audio chunk for a single sentence, in one voice (_id:false).
type Segment struct {
	SentenceID   bson.ObjectID `bson:"sentenceId"             json:"sentenceId"`
	AudioPath    *string       `bson:"audioPath,omitempty"    json:"audioPath,omitempty"`
	DurationSecs *float64      `bson:"durationSecs,omitempty" json:"durationSecs,omitempty"`
	AudioStatus  AudioStatus   `bson:"audioStatus"            json:"audioStatus"`
	AudioError   *string       `bson:"audioError,omitempty"   json:"audioError,omitempty"`
}

// VoiceTrack is a chapter's rendered audio for one voice (_id:false).
type VoiceTrack struct {
	Voice             string      `bson:"voice"                       json:"voice"`
	AudioPath         *string     `bson:"audioPath,omitempty"         json:"audioPath,omitempty"`
	AudioDurationSecs *float64    `bson:"audioDurationSecs,omitempty" json:"audioDurationSecs,omitempty"`
	AudioStatus       AudioStatus `bson:"audioStatus"                 json:"audioStatus"`
	AudioError        *string     `bson:"audioError,omitempty"        json:"audioError,omitempty"`
	Segments          []Segment   `bson:"segments"                    json:"segments"`
}

// Sentence is an editable, speech-ready sentence — the source of truth for
// chapter audio. `Text` is what's synthesized; `Display` is shown to the
// reader; `Original` is the pre-SLM-split source sentence.
type Sentence struct {
	ID       bson.ObjectID `bson:"_id"                json:"_id"`
	Order    int           `bson:"order"              json:"order"`
	Text     string        `bson:"text"               json:"text"`
	Display  *string       `bson:"display,omitempty"  json:"display,omitempty"`
	Original *string       `bson:"original,omitempty" json:"original,omitempty"`
}

type Chapter struct {
	ID        bson.ObjectID `bson:"_id"       json:"_id"`
	Title     string        `bson:"title"     json:"title"`
	StartPage int           `bson:"startPage" json:"startPage"`
	StartChar int           `bson:"startChar" json:"startChar"`
	Sentences []Sentence    `bson:"sentences" json:"sentences"`
	Tracks    []VoiceTrack  `bson:"tracks"    json:"tracks"`
}

type OcrPage struct {
	ID       bson.ObjectID `bson:"_id,omitempty"      json:"_id,omitempty"`
	Page     int           `bson:"page"               json:"page"`
	Text     string        `bson:"text"               json:"text"`
	ReadText *string       `bson:"readText,omitempty" json:"readText,omitempty"`
	Language string        `bson:"language"           json:"language"`
	Status   OcrStatus     `bson:"status"             json:"status"`
	Error    *string       `bson:"error,omitempty"    json:"error,omitempty"`
}

type Progress struct {
	Current int    `bson:"current" json:"current"`
	Total   int    `bson:"total"   json:"total"`
	Message string `bson:"message" json:"message"`
}

type Book struct {
	ID             bson.ObjectID `bson:"_id"                      json:"_id"`
	Name           string        `bson:"name"                     json:"name"`
	Status         BookStatus    `bson:"status"                   json:"status"`
	Language       *string       `bson:"language,omitempty"       json:"language,omitempty"`
	FolderPath     string        `bson:"folderPath"               json:"folderPath"`
	FilePath       string        `bson:"filePath"                 json:"filePath"`
	CoverImagePath *string       `bson:"coverImagePath,omitempty" json:"coverImagePath,omitempty"`
	SummaryPages   []int         `bson:"summaryPages"             json:"summaryPages"`
	CoverPage      int           `bson:"coverPage"                json:"coverPage"`
	FirstPage      int           `bson:"firstPage"                json:"firstPage"`
	LastPage       int           `bson:"lastPage"                 json:"lastPage"`
	TotalPages     int           `bson:"totalPages"               json:"totalPages"`
	Chapters       []Chapter     `bson:"chapters"                 json:"chapters"`
	OcrPages       []OcrPage     `bson:"ocrPages"                 json:"ocrPages"`
	Progress       Progress      `bson:"progress"                 json:"progress"`
	ErrorMessage   *string       `bson:"errorMessage,omitempty"   json:"errorMessage,omitempty"`
	Voices         []string      `bson:"voices"                   json:"voices"`
	Deleted        bool          `bson:"deleted"                  json:"deleted"`
	CreatedAt      DateTime      `bson:"createdAt"                json:"createdAt"`
	UpdatedAt      DateTime      `bson:"updatedAt"                json:"updatedAt"`
	V              int32         `bson:"__v"                      json:"__v"`
}

func (b *Book) TrackForVoice(chapter *Chapter, voice string) *VoiceTrack {
	for i := range chapter.Tracks {
		if chapter.Tracks[i].Voice == voice {
			return &chapter.Tracks[i]
		}
	}
	return nil
}

func FreshTracks(voices []string) []VoiceTrack {
	tracks := make([]VoiceTrack, len(voices))
	for i, voice := range voices {
		tracks[i] = VoiceTrack{Voice: voice, AudioStatus: AudioPending, Segments: []Segment{}}
	}
	return tracks
}

// DeriveTrackStatus computes a track's status from its segments (the
// assembled-chapter readiness).
func DeriveTrackStatus(segments []Segment) AudioStatus {
	if len(segments) == 0 {
		return AudioPending
	}
	counts := map[AudioStatus]int{}
	for _, s := range segments {
		counts[s.AudioStatus]++
	}
	switch {
	case counts[AudioGenerating] > 0:
		return AudioGenerating
	case counts[AudioError] > 0:
		return AudioError
	case counts[AudioStale] > 0:
		return AudioStale
	case counts[AudioComplete] == len(segments):
		return AudioComplete
	default:
		return AudioPending
	}
}

// ClientTrack / ClientChapter are the wire DTOs of serializeChaptersForClient:
// chapters trimmed for sync payloads — sentences and per-segment data stay
// server-side.
type ClientTrack struct {
	Voice             string      `json:"voice"`
	AudioPath         *string     `json:"audioPath,omitempty"`
	AudioDurationSecs *float64    `json:"audioDurationSecs,omitempty"`
	AudioStatus       AudioStatus `json:"audioStatus"`
	AudioError        *string     `json:"audioError,omitempty"`
}

type ClientChapter struct {
	ID        bson.ObjectID `json:"_id"`
	Title     string        `json:"title"`
	StartPage int           `json:"startPage"`
	StartChar int           `json:"startChar"`
	Tracks    []ClientTrack `json:"tracks"`
}

func SerializeChaptersForClient(chapters []Chapter) []ClientChapter {
	out := make([]ClientChapter, len(chapters))
	for i, c := range chapters {
		tracks := make([]ClientTrack, len(c.Tracks))
		for j, t := range c.Tracks {
			tracks[j] = ClientTrack{
				Voice:             t.Voice,
				AudioPath:         t.AudioPath,
				AudioDurationSecs: t.AudioDurationSecs,
				AudioStatus:       t.AudioStatus,
				AudioError:        t.AudioError,
			}
		}
		out[i] = ClientChapter{ID: c.ID, Title: c.Title, StartPage: c.StartPage, StartChar: c.StartChar, Tracks: tracks}
	}
	return out
}

// ClientBook is a Book with chapters trimmed and OCR text sanitized — the
// shape sent on the WS sync path (port of sanitizeBook in routes/books.ts).
type ClientBook struct {
	Book
	Chapters []ClientChapter `json:"chapters"`
}

func SanitizeBook(b *Book) ClientBook {
	out := ClientBook{Book: *b, Chapters: SerializeChaptersForClient(b.Chapters)}
	out.Book.Chapters = nil
	pages := make([]OcrPage, len(b.OcrPages))
	for i, p := range b.OcrPages {
		p.Text = sanitize.PageText(p.Text)
		pages[i] = p
	}
	out.Book.OcrPages = pages
	return out
}

// Normalize ensures every slice the frontend expects as [] is non-nil, so
// nil slices never encode as null in JSON or BSON (Mongoose always stores []).
func (b *Book) Normalize() {
	if b.SummaryPages == nil {
		b.SummaryPages = []int{}
	}
	if b.Voices == nil {
		b.Voices = []string{}
	}
	if b.Chapters == nil {
		b.Chapters = []Chapter{}
	}
	if b.OcrPages == nil {
		b.OcrPages = []OcrPage{}
	}
	for i := range b.Chapters {
		c := &b.Chapters[i]
		if c.Sentences == nil {
			c.Sentences = []Sentence{}
		}
		if c.Tracks == nil {
			c.Tracks = []VoiceTrack{}
		}
		for j := range c.Tracks {
			if c.Tracks[j].Segments == nil {
				c.Tracks[j].Segments = []Segment{}
			}
		}
	}
}
